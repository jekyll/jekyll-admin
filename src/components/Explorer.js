import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { browserHistory, Link } from 'react-router';
import DocumentTitle from 'react-document-title';
import Breadcrumbs from './Breadcrumbs';
import Button from './Button';
import InputSearch from './form/InputSearch';
import { ADMIN_PREFIX } from '../constants';
import { getDeleteMessage, getNotFoundMessage } from '../translations';

export default class Explorer extends Component {
  handleClickDelete(filename) {
    const { deleteItem, params, type } = this.props;
    const confirm = window.confirm(getDeleteMessage(filename));
    const dir = params.splat ? `/${params.splat}` : '';
    if (confirm) {
      deleteItem(params.splat, filename);
      browserHistory.push(`${ADMIN_PREFIX}/${type}${dir}`);
    }
  }

  renderTitleCell(name, icon, type, path) {
    return (
      <td className="row-title">
        <strong>
          <Link to={`${ADMIN_PREFIX}/${type}/${path}`}>
            <i className={`fa fa-${icon}`} aria-hidden="true" />
            {name}
          </Link>
        </strong>
      </td>
    );
  }

  renderDirectoryRow(directory, type, index) {
    const { name, relative_path } = directory;
    return (
      <tr key={index}>
        {this.renderTitleCell(name, 'folder', type, relative_path)}
        <td />
      </tr>
    );
  }

  renderFileRow(file, type, index) {
    const { name, relative_path, http_url } = file;
    return (
      <tr key={index}>
        {this.renderTitleCell(name, 'file-text-o', type, relative_path)}
        <td>
          <div className="row-actions">
            <Button
              onClick={() => this.handleClickDelete(name)}
              type="delete"
              icon="trash"
              active={true}
              thin
            />
            {http_url && (
              <Button to={http_url} type="view" icon="eye" active thin />
            )}
          </div>
        </td>
      </tr>
    );
  }

  renderRows() {
    const { items, type } = this.props;
    return items.map((entry, index) =>
      entry.type == 'directory'
        ? this.renderDirectoryRow(entry, type, index)
        : this.renderFileRow(entry, type, index)
    );
  }

  render() {
    const { items, type, newBtnLabel, search, params } = this.props;
    const typeLabel = type === 'datafiles' ? 'data files' : type;

    const to = params.splat
      ? `${ADMIN_PREFIX}/${type}/${params.splat}/new`
      : `${ADMIN_PREFIX}/${type}/new`;

    const content = items.length ? (
      <div className="content-table">
        <table>
          <thead>
            <tr>
              <th>Filename</th>
              <th className="th-actions">Actions</th>
            </tr>
          </thead>
          <tbody>{this.renderRows()}</tbody>
        </table>
      </div>
    ) : (
      <h1>{getNotFoundMessage(typeLabel)}</h1>
    );

    return (
      <div>
        <div className="content-header">
          <Breadcrumbs type={type} splat={params.splat} />
          <div className="page-buttons">
            <Link className="btn btn-active" to={to}>
              {newBtnLabel}
            </Link>
          </div>
          <div className="pull-right">
            <InputSearch searchBy="filename" search={search} />
          </div>
        </div>
        {content}
      </div>
    );
  }
}

Explorer.propTypes = {
  type: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
  params: PropTypes.object.isRequired,
  newBtnLabel: PropTypes.string.isRequired,
  deleteItem: PropTypes.func.isRequired,
  search: PropTypes.func.isRequired,
};
