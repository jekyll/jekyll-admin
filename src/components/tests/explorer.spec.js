import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Explorer from '../Explorer';
import { resources } from './fixtures';

const defaultProps = {
  type: 'resources',
  items: resources,
  params: {},
  newBtnLabel: 'New resource',
};

const actions = {
  deleteItem: jest.fn(),
  search: jest.fn(),
};

function setup(overrides = {}) {
  const component = mount(
    <Explorer {...defaultProps} {...actions} {...overrides} />
  );

  return {
    component,
    actions,
    header: component.find('.content-header'),
  };
}

describe('Components::Explorer', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Explorer {...defaultProps} {...actions} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders breadcrumbs correctly', () => {
    const { header } = setup({
      type: 'datafiles',
      params: { splat: 'movies' },
    });
    const crumbs = header.find('li');
    expect(crumbs.map(c => c.text()).join(' / ')).toBe('Data Files / movies');
  });
});
