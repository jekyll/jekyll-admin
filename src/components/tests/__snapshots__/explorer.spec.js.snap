// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::Explorer renders correctly 1`] = `
<div>
  <div
    className="content-header"
  >
    <ul
      className="breadcrumbs"
    >
      <li>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          Resources
        </a>
      </li>
    </ul>
    <div
      className="page-buttons"
    >
      <a
        className="btn btn-active"
        onClick={[Function]}
        style={Object {}}
      >
        New resource
      </a>
    </div>
    <div
      className="pull-right"
    >
      <input
        className="field"
        onKeyPress={[Function]}
        placeholder="Search by filename"
        type="text"
      />
    </div>
  </div>
  <div
    className="content-table"
  >
    <table>
      <thead>
        <tr>
          <th>
            Filename
          </th>
          <th
            className="th-actions"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="row-title"
          >
            <strong>
              <a
                onClick={[Function]}
                style={Object {}}
              >
                <i
                  aria-hidden="true"
                  className="fa fa-folder"
                />
                resource-dir
              </a>
            </strong>
          </td>
          <td />
        </tr>
        <tr>
          <td
            className="row-title"
          >
            <strong>
              <a
                onClick={[Function]}
                style={Object {}}
              >
                <i
                  aria-hidden="true"
                  className="fa fa-file-text-o"
                />
                resource.md
              </a>
            </strong>
          </td>
          <td>
            <div
              className="row-actions"
            >
              <button
                className="btn btn-active btn-delete btn-thin"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="fa fa-trash"
                />
                Delete
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
